src/io/AtomicRelaxationReader.cc

```diff
@@ -0,0 +1,172 @@
+//---------------------------------------------------------------------------//
+#include "AtomicRelaxationReader.hh"
+
+#include <fstream>
+#include <sstream>
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the reader using the G4LEDATA environment variable to get the path
+ * to the data.
+ */
+AtomicRelaxationReader::AtomicRelaxationReader()
+{
+    const char* env_var = std::getenv("G4LEDATA");
+    CELER_VALIDATE(env_var, "Environment variable G4LEDATA is not defined.");
+    {
+        std::ostringstream os;
+        os << env_var << "/fluor";
+        fluor_path_ = os.str();
+    }
+    {
+        std::ostringstream os;
+        os << env_var << "/auger";
+        auger_path_ = os.str();
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the reader with the path to the directory containing the data.
+ */
+AtomicRelaxationReader::AtomicRelaxationReader(const char* fluor_path,
+                                               const char* auger_path)
+    : fluor_path_(fluor_path), auger_path_(auger_path)
+{
+    CELER_EXPECT(!fluor_path_.empty());
+    CELER_EXPECT(!auger_path_.empty());
+
+    if (fluor_path_.back() == '/')
+        fluor_path_.pop_back();
+    if (auger_path_.back() == '/')
+        auger_path_.pop_back();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Read the data for the given element.
+ */
+AtomicRelaxationReader::result_type
+AtomicRelaxationReader::operator()(int atomic_number) const
+{
+    CELER_EXPECT(atomic_number > 0 && atomic_number < 101);
+
+    // EADL does not provide transition data for Z < 6
+    result_type result;
+    if (atomic_number < 6)
+    {
+        return result;
+    }
+
+    std::string Z = std::to_string(atomic_number);
+
+    // Read fluorescence transition probabilities and subshell designators. All
+    // lines in the data file are 3 columns. The first line of each section is
+    // the subshell designator of the initial vacancy. Each of the following
+    // rows contains data for a different transition, where the columns are:
+    // designator of the shell of the electron that fills the vacancy,
+    // transition probability, and transition energy. A row of -1 marks the end
+    // of the shell data, and -2 marks the end of the file.
+    {
+        std::string   filename = fluor_path_ + "/fl-tr-pr-" + Z + ".dat";
+        std::ifstream infile(filename);
+        CELER_VALIDATE(infile, "Couldn't open '" << filename << "'");
+
+        int       des       = 0;
+        real_type tr_energy = 0.;
+        real_type tr_prob   = 0.;
+        infile >> des >> des >> des;
+        CELER_ASSERT(infile);
+        result.designators.push_back(des);
+        result.fluor.emplace_back();
+        while (infile >> des >> tr_prob >> tr_energy)
+        {
+            CELER_ASSERT(infile);
```

Comment:
Oh duh, thanks.
1. You're right, it makes much more sense for each `SubshellInput` to be a vector of `struct TransitionInput`, etc.,  and similarly in the interface. With the SoA I wasn't allocating storage for the Auger shell IDs in the fluorescence data when Auger was turned off, but it's a really small amount of data, and doing it this way will simplify things significantly in the `Params` constructor as well as here in the reader. Liking this much better.
2. Definitely, good point.
3. > If the first values of des >> des >> des are giving the size of the file/number of designators (which is what it looks like for the auger data), maybe it's better to restructure it as a for loop?

   Unfortunately no, we don't get the size/number of designators -- that's just the subshell designator for the first section of transition data. Let me see if I can structure things a little more cleanly, though.. don't worry, any time I spend on this will be less than the time I've spent going over the Geant4 input processing 😉

Reply:
You're a trooper 😄 Don't worry about the reader implementation if it works. I think we have enough other stuff that would be a better sink for our time.